### ISSUE TITLE
"Check failed" with multiple lax.conv_general_dilated, vmap and jit

### ISSUE THREAD

**Issue by tristandeleu** [google/jax#4819]
I have come across the following issue:
```
F external/org_tensorflow/tensorflow/compiler/xla/service/gpu/gpu_conv_rewriter.cc:543] Check failed: std::abs(input_feature_dimension - output_feature_dimension) == 1LL (3 vs. 1)
```

In my case, I was trying to implement MAML (very loosely following the [notebook available in the docs](https://github.com/google/jax/blob/8f1be202a35639cabc41d916d78f3e6404e52b72/docs/notebooks/maml.ipynb)) for image classification. The issue I am having appears when I use `jit` (on GPU) on a loss function with multiple convolutions (from `lax.conv_general_dilated`) and `vmap` to apply the function to a batch of tasks. Here is a minimal reproducible example of this issue:

```python
import jax
import jax.numpy as jnp

from jax import vmap, lax, random, grad, tree_util
from functools import partial


def loss_fn(params, inputs, targets):
    """Loss function"""
    # Unpack parameters
    params1, params2 = params
    dimension_numbers = lax.ConvDimensionNumbers(
        lhs_spec=(0, 3, 1, 2),  # Channel last
        rhs_spec=(3, 2, 0, 1),
        out_spec=(0, 3, 1, 2)
    )
    
    # First convolution
    outputs = lax.conv_general_dilated(inputs, params1,
        window_strides=(1, 1), padding='SAME',
        lhs_dilation=None, rhs_dilation=None,
        dimension_numbers=dimension_numbers, feature_group_count=1)
    
    # Second covolution
    outputs = lax.conv_general_dilated(outputs, params2,
        window_strides=(1, 1), padding='SAME',
        lhs_dilation=None, rhs_dilation=None,
        dimension_numbers=dimension_numbers, feature_group_count=1)
    
    # MSE loss
    return jnp.mean((outputs - targets) ** 2)

def adaptation(params, inputs, targets):
    """MAML: adaptation with one-step gradient descent"""
    grads = grad(loss_fn)(params, inputs, targets)
    return tree_util.tree_multimap(lambda p, g: p - 0.1 * g, params, grads)

def _outer_loss_fn(params, train_inputs, train_targets, test_inputs, test_targets):
    """Apply outer-loss to a single task"""
    adapted_params = adaptation(params, train_inputs, train_targets)
    return loss_fn(adapted_params, test_inputs, test_targets)

def outer_loss_fn(params, train_inputs, train_targets, test_inputs, test_targets):
    """Outer-loss on a batch of tasks"""
    losses = vmap(partial(_outer_loss_fn, params))(
        train_inputs, train_targets, test_inputs, test_targets)
    return jnp.mean(losses)

step = grad(outer_loss_fn)

# Inputs / targets
# Shape: (task batch size, number of examples per task, height, width, channels)
# The number of examples per task can be different between support / query datasets
train_inputs = jnp.zeros((3, 5, 28, 28, 1))
train_targets = jnp.zeros((3, 5, 28, 28, 32))

test_inputs = jnp.zeros((3, 11, 28, 28, 1))
test_targets = jnp.zeros((3, 11, 28, 28, 32))

# Initialize the parameters
params1 = jnp.zeros((3, 3, 1, 32))
params2 = jnp.zeros((3, 3, 32, 32))
params = (params1, params2)

# Error
jax.jit(step)(params, train_inputs, train_targets, test_inputs, test_targets)
# F external/org_tensorflow/tensorflow/compiler/xla/service/gpu/gpu_conv_rewriter.cc:543]
# Check failed: std::abs(input_feature_dimension - output_feature_dimension) == 1LL (3 vs. 1)
```

Some cases where there is no error:
 - If there is no `jit`
<details>
<summary>Expand</summary>

```python
import jax
import jax.numpy as jnp

from jax import vmap, lax, random, grad, tree_util
from functools import partial


def loss_fn(params, inputs, targets):
    """Loss function"""
    # Unpack parameters
    params1, params2 = params
    dimension_numbers = lax.ConvDimensionNumbers(
        lhs_spec=(0, 3, 1, 2),  # Channel last
        rhs_spec=(3, 2, 0, 1),
        out_spec=(0, 3, 1, 2)
    )
    
    # First convolution
    outputs = lax.conv_general_dilated(inputs, params1,
        window_strides=(1, 1), padding='SAME',
        lhs_dilation=None, rhs_dilation=None,
        dimension_numbers=dimension_numbers, feature_group_count=1)
    
    # Second covolution
    outputs = lax.conv_general_dilated(outputs, params2,
        window_strides=(1, 1), padding='SAME',
        lhs_dilation=None, rhs_dilation=None,
        dimension_numbers=dimension_numbers, feature_group_count=1)
    
    # MSE loss
    return jnp.mean((outputs - targets) ** 2)

def adaptation(params, inputs, targets):
    """MAML: adaptation with one-step gradient descent"""
    grads = grad(loss_fn)(params, inputs, targets)
    return tree_util.tree_multimap(lambda p, g: p - 0.1 * g, params, grads)

def _outer_loss_fn(params, train_inputs, train_targets, test_inputs, test_targets):
    """Apply outer-loss to a single task"""
    adapted_params = adaptation(params, train_inputs, train_targets)
    return loss_fn(adapted_params, test_inputs, test_targets)

def outer_loss_fn(params, train_inputs, train_targets, test_inputs, test_targets):
    """Outer-loss on a batch of tasks"""
    losses = vmap(partial(_outer_loss_fn, params))(
        train_inputs, train_targets, test_inputs, test_targets)
    return jnp.mean(losses)

step = grad(outer_loss_fn)

# Inputs / targets
# Shape: (task batch size, number of examples per task, height, width, channels)
# The number of examples per task can be different between support / query datasets
train_inputs = jnp.zeros((3, 5, 28, 28, 1))
train_targets = jnp.zeros((3, 5, 28, 28, 32))

test_inputs = jnp.zeros((3, 11, 28, 28, 1))
test_targets = jnp.zeros((3, 11, 28, 28, 32))

# Initialize the parameters
params1 = jnp.zeros((3, 3, 1, 32))
params2 = jnp.zeros((3, 3, 32, 32))
params = (params1, params2)

# No Error
step(params, train_inputs, train_targets, test_inputs, test_targets)
```
</details>

 - If there is only a single convolution
<details>
<summary>Expand</summary>

```python
import jax
import jax.numpy as jnp

from jax import vmap, lax, random, grad, tree_util
from functools import partial


def loss_fn(params, inputs, targets):
    """Loss function"""
    dimension_numbers = lax.ConvDimensionNumbers(
        lhs_spec=(0, 3, 1, 2),  # Channel last
        rhs_spec=(3, 2, 0, 1),
        out_spec=(0, 3, 1, 2)
    )

    # Second covolution
    outputs = lax.conv_general_dilated(inputs, params,
        window_strides=(1, 1), padding='SAME',
        lhs_dilation=None, rhs_dilation=None,
        dimension_numbers=dimension_numbers, feature_group_count=1)

    # MSE loss
    return jnp.mean((outputs - targets) ** 2)

def adaptation(params, inputs, targets):
    """MAML: adaptation with one-step gradient descent"""
    grads = grad(loss_fn)(params, inputs, targets)
    return tree_util.tree_multimap(lambda p, g: p - 0.1 * g, params, grads)

def _outer_loss_fn(params, train_inputs, train_targets, test_inputs, test_targets):
    """Apply outer-loss to a single task"""
    adapted_params = adaptation(params, train_inputs, train_targets)
    return loss_fn(adapted_params, test_inputs, test_targets)

def outer_loss_fn(params, train_inputs, train_targets, test_inputs, test_targets):
    """Outer-loss on a batch of tasks"""
    losses = vmap(partial(_outer_loss_fn, params))(
        train_inputs, train_targets, test_inputs, test_targets)
    return jnp.mean(losses)

step = grad(outer_loss_fn)

# Inputs / targets
# Shape: (task batch size, number of examples per task, height, width, channels)
# The number of examples per task can be different between support / query datasets
train_inputs = jnp.zeros((3, 5, 28, 28, 1))
train_targets = jnp.zeros((3, 5, 28, 28, 32))

test_inputs = jnp.zeros((3, 11, 28, 28, 1))
test_targets = jnp.zeros((3, 11, 28, 28, 32))

# Initialize the parameters
params = jnp.zeros((3, 3, 1, 32))

# No Error
jax.jit(step)(params, train_inputs, train_targets, test_inputs, test_targets)
```
</details>

 - If there is no `vmap` (only applying the outer-loss to a single task)
<details>
<summary>Expand</summary>

```python
import jax
import jax.numpy as jnp

from jax import vmap, lax, random, grad, tree_util
from functools import partial


def loss_fn(params, inputs, targets):
    """Loss function"""
    # Unpack parameters
    params1, params2 = params
    dimension_numbers = lax.ConvDimensionNumbers(
        lhs_spec=(0, 3, 1, 2),  # Channel last
        rhs_spec=(3, 2, 0, 1),
        out_spec=(0, 3, 1, 2)
    )

    # First convolution
    outputs = lax.conv_general_dilated(inputs, params1,
        window_strides=(1, 1), padding='SAME',
        lhs_dilation=None, rhs_dilation=None,
        dimension_numbers=dimension_numbers, feature_group_count=1)

    # Second covolution
    outputs = lax.conv_general_dilated(outputs, params2,
        window_strides=(1, 1), padding='SAME',
        lhs_dilation=None, rhs_dilation=None,
        dimension_numbers=dimension_numbers, feature_group_count=1)

    # MSE loss
    return jnp.mean((outputs - targets) ** 2)

def adaptation(params, inputs, targets):
    """MAML: adaptation with one-step gradient descent"""
    grads = grad(loss_fn)(params, inputs, targets)
    return tree_util.tree_multimap(lambda p, g: p - 0.1 * g, params, grads)

def _outer_loss_fn(params, train_inputs, train_targets, test_inputs, test_targets):
    """Apply outer-loss to a single task"""
    adapted_params = adaptation(params, train_inputs, train_targets)
    return loss_fn(adapted_params, test_inputs, test_targets)

step = grad(_outer_loss_fn)

key = random.PRNGKey(0)

# Inputs / targets
# Shape: (task batch size, number of examples per task, height, width, channels)
# The number of examples per task can be different between support / query datasets
train_inputs = jnp.zeros((3, 5, 28, 28, 1))
train_targets = jnp.zeros((3, 5, 28, 28, 32))

test_inputs = jnp.zeros((3, 11, 28, 28, 1))
test_targets = jnp.zeros((3, 11, 28, 28, 32))

# Initialize the parameters
params1 = jnp.zeros((3, 3, 1, 32))
params2 = jnp.zeros((3, 3, 32, 32))
params = (params1, params2)

# No error
jax.jit(step)(params, train_inputs[0], train_targets[0], test_inputs[0], test_targets[0])
```
</details>

Here are the versions I am using
 - `jax==0.2.6`
 - `jaxlib==0.1.56`
 - CUDA version: 10.1
 - CUDNN version: 7.3

**Comment by gnecula** [comment]
I tried to reproduce this on the HEAD version inside google, and I could not. Then I saw that the assert that fails has been changed in [XLA on October 26](https://github.com/tensorflow/tensorflow/commit/edfc5938ba99cbe81ac50796f6ff647a374daf82). I think that this will likely fix your problem. (the jaxlib has last been released on October 14th).

Would you be able to try to check if the bug goes away if you build [jaxlib from source](https://jax.readthedocs.io/en/latest/developer.html#building-jaxlib-from-source)?

**Comment by tristandeleu** [comment]
Thank you! It does work as expected when I build jaxlib from source (on 3e7285b).
